File: alternative/medium/letter_combinations_of_a_phone_number.py
from typing import List

tn_representations = {
    "2": "abc",
    "3": "def",
    "4": "ghi",
    "5": "jkl",
    "6": "mon",
    "7": "pqrs",
    "8": "tuv",
    "9": "wxyz"
}

def letterCombinations(digits: str) -> List[str]:
    res = []

    if len(digits) == 0:
        return res

    def backtracking(i, curStr):
        if len(curStr) == len(digits):
            res.append(curStr)
            return
        
        for c in tn_representations[digits[i]]:
            backtracking(i+1, curStr + c)


    backtracking(0, "")
    return res


assert letterCombinations("23") == ["ad","ae","af","bd","be","bf","cd","ce","cf"]

File: src/medium/letter_combinations_of_a_phone_number.rs
#![allow(dead_code)]
pub fn letter_combinations(digits: String) -> Vec<String> {
    if digits.is_empty() {
        return vec![];
    }

    let phone_map = vec!["abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"];
    let mut output = Vec::new();

    fn backtrack(
        combination: String,
        next_digits: &str,
        phone_map: &Vec<&str>,
        output: &mut Vec<String>,
    ) {
        if next_digits.is_empty() {
            output.push(combination);
        } else {
            let letters = phone_map[next_digits.chars().nth(0).unwrap() as usize - '2' as usize];
            for letter in letters.chars() {
                let new_combination = combination.clone() + &letter.to_string();
                backtrack(new_combination, &next_digits[1..], phone_map, output);
            }
        }
    }

    backtrack(String::new(), &digits, &phone_map, &mut output);
    output
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_letter_combinations() {
        let result = letter_combinations("23".to_string());
        let expected = vec![
            "ad".to_string(),
            "ae".to_string(),
            "af".to_string(),
            "bd".to_string(),
            "be".to_string(),
            "bf".to_string(),
            "cd".to_string(),
            "ce".to_string(),
            "cf".to_string(),
        ];
        assert_eq!(result, expected);
    }
}
